common/map_coordinate.h

```diff
@@ -0,0 +1,231 @@
+// OSGB-36 datum - used for UK OS map coordinates
+struct OSGB36
+{
+    static const Transform transform;
+    static const Ellipsoid ellipsoid;
+};
+const Transform OSGB36::transform{units::length::meter_t(-446.448), units::length::meter_t(125.157), units::length::meter_t(-542.060),
+                                  20.4894, units::angle::arcsecond_t(-0.1502), units::angle::arcsecond_t(-0.2470), units::angle::arcsecond_t(-0.8421)};
+const Ellipsoid OSGB36::ellipsoid{units::length::meter_t(6377563.396), units::length::meter_t(6356256.909), 1.0 / 299.3249646};
+
+// An OS coordinate consisting of an 'easting' and 'northing' distance
+typedef std::tuple<units::length::meter_t, units::length::meter_t> OSCoordinate;
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Ellipsoid
+//----------------------------------------------------------------------------
+//! A latitude and longitude relative to a particular datum
+template<typename D>
+struct LatLon
+{
+    typedef D Datum;
+
+    const units::angle::degree_t lat;
+    const units::angle::degree_t lon;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Cartesian
+//----------------------------------------------------------------------------
+//! A cartesian coordinate relative to a particular datum
+template<typename D>
+struct Cartesian
+{
+    typedef D Datum;
+
+    const units::length::meter_t x;
+    const units::length::meter_t y;
+    const units::length::meter_t z;
+};
+
+
+// Convert latitude and longitude to cartesian
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+Cartesian<Datum> latLonToCartesian(const LatLon<Datum> &latLon)
+{
+    using namespace units::math;
+    using namespace units::length;
+    using namespace units::literals;
+
+    const meter_t h = 0.0_m; // height above ellipsoid - not currently used
+    const double sinPhi = sin(latLon.lat);
+    const double cosPhi = cos(latLon.lat);
+    const double sinLambda = sin(latLon.lon);
+    const double cosLambda = cos(latLon.lon);
+
+    const double eSq = 2.0 * Datum::ellipsoid.f - Datum::ellipsoid.f * Datum::ellipsoid.f;  // 1st eccentricity squared ≡ (a²-b²)/a²
+    const meter_t nu = Datum::ellipsoid.a / std::sqrt(1.0 - eSq * sinPhi * sinPhi);         // radius of curvature in prime vertical
+
+    return Cartesian<Datum>{(nu + h) * cosPhi * cosLambda,
+                            (nu + h) * cosPhi * sinLambda,
+                            (nu * (1.0 - eSq) + h) * sinPhi};
+}
+
+// Transform cartesian coordinates from WGS84 into desired space
+// Ported from Javascript Ordnance Survey Grid Reference functions (c) Chris Veness 2005-2017 (MIT Licence)
+template<typename Datum>
+Cartesian<Datum> transformCartesian(const Cartesian<WGS84> &c)
```

Comment:
Is this function specific to the WGS84 datum? Or could we just write it as a template functoin with ``Datum1`` and ``Datum2``?

Reply:
Basically yes - all the transforms are transforming from WGS84 (maybe I should make that clearer in the name of the transforms)

Reply:
But what I meant was would this function also work for tranforming e.g. from something else to WGS84? In which case it might be worth having the other datum also as a template parameter just to indiciate that you're not doing any weird WGS84-specific things, and also in case we decide to convert from something other than WGS84 in future.

Reply:
No, it only transforms from cartesian coordinates in the WGS84 datum to another datum of your choice

Reply:
...another thought. You could have a ``.transform()`` member function of ``Cartesian`` instead of having it as a separate function. Just a thought though.

Reply:
Ahh fair enough. Dw about it then.